### ISSUE TITLE
Issues and suggestions related to performance

### ISSUE THREAD

**Issue by forthrin** [yt-dlp/yt-dlp#5829]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a feature unrelated to a specific site
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I'm running yt-dlp version **2022.11.11** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

Getting the final URL for sites usually takes several seconds, whereas the same can often be accomplished with a small, customized script in a fraction of the time. So I did some benchmarking to look into this. See results below.

Questions:

1. Are there plans to optimize away the general overhead that adds seconds to the performance?
2. Why are multiple m3u8/MPD files loaded? Can this be optimized? Is there a setting to get only one?
3. To put focus on performance, would it be possible to add a CLI option to time program execution (like behow)?

Thanks in advance for your kind response and for making this improved and well-maintained fork!

```
$ sudo purge && yt-dlp -v -g https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul
0.08 0.08 [debug] Command-line config: ['-v', '-g', 'https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul']
---> WHAT TAKES A SECOND HERE?
0.92 1.00 [debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
0.00 1.00 [debug] yt-dlp version 2022.11.11 [8b644025b] (source)
0.00 1.00 [debug] Lazy loading extractors is disabled
0.00 1.00 [debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
0.09 1.09 [debug] Git HEAD: 5424dba
0.07 1.16 [debug] Python 3.10.9 (CPython x86_64 64bit) - macOS-12.6.2-x86_64-i386-64bit (OpenSSL 1.1.1s  1 Nov 2022)
---> WHAT TAKES A SECOND HERE?
0.93 2.09 [debug] exe versions: ffmpeg git-2022-06-01-c6364b71 (setts), ffprobe git-2022-06-01-c6364b71, rtmpdump 2.4
0.00 2.10 [debug] Optional libraries: Cryptodome-3.15.0, certifi-2022.06.15, sqlite3-2.6.0, xattr-0.9.9
0.03 2.12 [debug] Proxy map: {}
0.08 2.20 [debug] Loaded 1736 extractors
---> WHAT TAKES HALF A SECOND HERE?
0.64 2.84 [debug] Using fake IP 78.72.0.76 (SE) as X-Forwarded-For
0.00 2.84 [SVTPlay] Extracting URL: https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul
0.00 2.84 [SVTPlay] 8MMzVRJ: Downloading webpage
0.11 2.95 [SVTPlay] 8MMzVRJ: Downloading JSON metadata
---> WHY ARE ALL THESE LOADED?
0.02 2.98 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.18 3.15 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.14 3.30 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.14 3.44 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.19 3.63 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.17 3.80 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.14 3.95 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.21 4.15 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.14 4.29 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.14 4.44 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.15 4.59 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.17 4.75 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.19 4.95 [debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
0.13 5.08 [debug] Default format spec: bestvideo*+bestaudio/best
0.01 5.08 [info] 8MMzVRJ: Downloading 1 format(s): dash-hbbtv-hevc-0+dash-full-13
0.06 5.14 https://svt-vod-6m.akamaized.net/d0/world/20221202/ec9513e7-343e-4af0-93f7-cad3ee8e679e/dash-hbbtv-hevc.mpd
```

Crude benchmarking code used:

```
diff --git a/yt_dlp/utils.py b/yt_dlp/utils.py
index 65408bf..cc9c3f1 100644
--- a/yt_dlp/utils.py
+++ b/yt_dlp/utils.py
@@ -2037,2 +2037,6 @@ def get_windows_version():
 
+import time
+start_time = time.time()
+last_time = start_time
+
 def write_string(s, out=None, encoding=None):
@@ -2051,2 +2055,7 @@ def write_string(s, out=None, encoding=None):
 
+    global last_time
+    curr_time = time.time()
+    diff = curr_time - last_time
+    last_time = curr_time
+    s = "\033[101m" if diff >= 0.4 else "\033[2m" + "{:.2f}\033[m \033[93m{:.2f}\033[m".format(diff, curr_time-start_time) + ' ' + s
     buffer.write(s.encode(enc, 'ignore') if enc else s)
```

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

**Comment by pukkandan** [comment]
> Are there plans to optimize away the general overhead that adds seconds to the performance?

It'll have to be decided on a case by case basis depending on how much benefit we get and any compromises that'll have to be made

> To put focus on performance, would it be possible to add a CLI option to time program execution (like behow)?

Use tools like `ts` to get timestamps on output. Or better, use Python profiling tools like cProfile

> 0.07 1.16 [debug] Python 3.10.9 (CPython x86_64 64bit) - macOS-12.6.2-x86_64-i386-64bit (OpenSSL 1.1.1s  1 Nov 2022)
> ---> WHAT TAKES A SECOND HERE?


When using `-v`, yt-dlp looks for available version of all dependencies to display in the header. This delay should disappear when not using `-v`

> 0.08 2.20 [debug] Loaded 1736 extractors
> ---> WHAT TAKES HALF A SECOND HERE?

Probably matching the correct extractor. This can be sped up by manually specifying `--use-extractor SVTPlay`

---

> Why are multiple m3u8/MPD files loaded? Can this be optimized? Is there a setting to get only one?
> ---> WHY ARE ALL THESE LOADED?


The site is providing multiple formats in multiple manifests. In general, we need to extract info about all of them before the best/requested formats can be determined. Relevant code (note the for loop) https://github.com/yt-dlp/yt-dlp/blob/c73355510629e3eda5a79d4e2876a35316ca6ed2/yt_dlp/extractor/svt.py#L24-L48

It may be possible to optimize out some of this - but hard to say without studying the extractor in detail. If it cannot be done, we could still provide some extractor-args to control this behavior

---

> 0.08 0.08 [debug] Command-line config: ['-v', '-g', 'https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul']
> ---> WHAT TAKES A SECOND HERE?

Command line option parsing is taking place here. No idea why it's taking so much time in your case. A cProfile output might be helpful (`python -m cProfile yt_dlp\__main__.py`)

> [debug] Lazy loading extractors is disabled

Actually, extractor imports also happen at this point. So lack of lazy loading could also be a factor for the slowdown. This can be enabled using `make lazy-extractors` (Remember to delete/rebuild this if u r doing extractor development)

Since dealing with m3u8 and this are completely separate matters, I propose we convert this issue into a "site enhancement" request to track svtplay specific improvements that can be made. You can open a new issue with the cProfile output so we can study what is slowing down the command line parsing

---

For reference, these are the times on my system

```
> ./yt-dlp.sh --ignore-config -v -g https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul &| ts -i %.s
0.307314 [debug] Command-line config: ['--ignore-config', '-v', '-g', 'https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul']
0.449267 [debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8 (No ANSI), error utf-8 (No ANSI), screen utf-8 (No ANSI)
0.000614 [debug] yt-dlp version 2022.11.11 [8b644025b] (source)
0.000404 [debug] Lazy loading extractors is disabled
0.000438 [debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
0.019340 [debug] Git HEAD: cf7809fb8
0.034558 [debug] Python 3.8.10 (CPython x86_64 64bit) - Linux-4.4.0-22000-Microsoft-x86_64-with-glibc2.29 (OpenSSL 1.1.1f  31 Mar 2020, glibc 2.31)
0.256557 [debug] exe versions: ffmpeg 4.2.7, ffprobe 4.2.7, rtmpdump 2.4
0.000602 [debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2019.11.28, mutagen-1.45.1, pyxattr-0.6.1, secretstorage-2.3.1, sqlite3-2.6.0, websockets-10.3
0.009968 [debug] Proxy map: {}
0.053326 [debug] Loaded 1736 extractors
0.361503 [debug] Using fake IP 78.72.200.216 (SE) as X-Forwarded-For
0.000429 [SVTPlay] Extracting URL: https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul
0.000384 [SVTPlay] 8MMzVRJ: Downloading webpage
0.156554 [SVTPlay] 8MMzVRJ: Downloading JSON metadata
0.091181 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.165694 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.120155 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.131730 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.155851 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.161911 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.102122 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.171295 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.106919 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.120786 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.105399 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.145176 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.136179 [debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
0.076718 [debug] Default format spec: bestvideo*+bestaudio/best
0.003679 [info] 8MMzVRJ: Downloading 1 format(s): dash-hbbtv-hevc-0+dash-full-13
0.028018 https://svt-vod-6m.akamaized.net/d0/world/20221202/ec9513e7-343e-4af0-93f7-cad3ee8e679e/dash-hbbtv-hevc.mpd
0.000492 https://svt-vod-6m.akamaized.net/d0/world/20221202/ec9513e7-343e-4af0-93f7-cad3ee8e679e/dash-full.mpd
```

Without `-v`

```
> ./yt-dlp.sh --ignore-config -s https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul &| ts -i %.s
1.186749 [SVTPlay] Extracting URL: https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul
0.000779 [SVTPlay] 8MMzVRJ: Downloading webpage
0.154440 [SVTPlay] 8MMzVRJ: Downloading JSON metadata
0.116872 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.138380 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.142884 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.128202 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.147723 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.146852 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.135264 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.179615 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.139817 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.104468 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.146493 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.133286 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.228905 [info] 8MMzVRJ: Downloading 1 format(s): dash-hbbtv-hevc-0+dash-full-13
```

**Comment by forthrin** [comment]
Thanks for your prompt and helpful feedback. Great tip about `ts`in `moreutils`!

- Skipping `-v` saves 1 second.
- Adding `--use-extractor SVTPlay` saves 0.5 seconds.
- Using `make lazy-extractors` seems to save nothing.

3.8 seconds is still the fastest I get with `yt-dlp´ (without --verbose).

[cProfile.txt](https://github.com/yt-dlp/yt-dlp/files/10258051/cProfile.txt)

PS! Can you **purge your disk cache** before running tests on your end, so that doesn't skew the results?

PPS! Let's look at this manually. Getting the page takes half a second:

```
$ time curl https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul
0.5s
```

However, it seems most links to single videos on this particular site contain the ID in the URL, so then we can skip fetching the HTML page, and get the JSON data directly, which is super-fast:

```
$ time curl https://api.svt.se/video/8MMzVRJ | jq -r '.videoReferences[] | [.format, .url] | @tsv'
0.06s
```

Now, given proper knowledge of the `format` keywords, we should be able to narrow down to the single media file the user is interested in (or is considered most valuable by default), rather than getting all of them.

In effect it is possible to start playing a given video almost instantly, as opposed to waiting 4 seconds.

Now, I realise that `yt-dlp` is a large utility, but growth in features and flexibility should't bloat the general performance of simple tasks in a given piece of software. (At least there is no splash screen yet ;))

Arguably the only thing that should really take time is the network traffic. 

Looking forward to your response on how to get rid of bottlenecks and get great performance!

```
$ sudo purge && time ./yt-dlp.sh --ignore-config --verbose --use-extractor SVTPlay -g https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul 2>&1 | ts -i %.s
>>1.058672 [debug] Command-line config: ['--ignore-config', '--verbose', '--use-extractor', 'SVTPlay', '-g', 'https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul']
0.043360 [debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8 (No ANSI), error utf-8 (No ANSI), screen utf-8 (No ANSI)
0.000100 [debug] yt-dlp version 2022.11.11 [8b644025b] (source)
0.000037 [debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
0.082500 [debug] Git HEAD: 5424dba
0.068977 [debug] Python 3.10.9 (CPython x86_64 64bit) - macOS-12.6.2-x86_64-i386-64bit (OpenSSL 1.1.1s  1 Nov 2022)
>>0.825499 [debug] exe versions: ffmpeg git-2022-06-01-c6364b71 (setts), ffprobe git-2022-06-01-c6364b71, rtmpdump 2.4
0.000730 [debug] Optional libraries: Cryptodome-3.15.0, certifi-2022.06.15, sqlite3-2.6.0, xattr-0.9.9
0.034857 [debug] Proxy map: {}
0.005168 [debug] Loaded 1 extractors
0.005077 [debug] Using fake IP 78.66.31.182 (SE) as X-Forwarded-For
0.000107 [SVTPlay] Extracting URL: https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul
0.000064 [SVTPlay] 8MMzVRJ: Downloading webpage
0.114435 [SVTPlay] 8MMzVRJ: Downloading JSON metadata
0.021117 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.180160 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.143044 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.140735 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.188321 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.168271 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.142624 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.243266 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.141948 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.146079 [SVTPlay] 8MMzVRJ: Downloading m3u8 information
0.146625 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.163831 [SVTPlay] 8MMzVRJ: Downloading MPD manifest
0.193227 [debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
0.120630 [debug] Default format spec: bestvideo*+bestaudio/best
0.005518 [info] 8MMzVRJ: Downloading 1 format(s): dash-hbbtv-hevc-0+dash-full-13
0.046443 https://svt-vod-6m.akamaized.net/d0/world/20221202/ec9513e7-343e-4af0-93f7-cad3ee8e679e/dash-hbbtv-hevc.mpd
0.000065 https://svt-vod-6m.akamaized.net/d0/world/20221202/ec9513e7-343e-4af0-93f7-cad3ee8e679e/dash-full.mpd
0m4.767s
```

**Comment by pukkandan** [comment]
> Now, given proper knowledge of the `format` keywords, we should be able to narrow down to the single media file the user is interested in (or is considered most valuable by default)

How?

> [cProfile.txt](https://github.com/yt-dlp/yt-dlp/files/10258051/cProfile.txt)
```
   cumtime   ncalls  tottime  percall  percall filename:lineno(function)
     0.648       31    0.001    0.000    0.021 __init__.py:1(<module>)
```

The whole startup took only 0.6s this time, not 2.2s like your original run. This much is to be expected since this is a large python program

> PS! Can you purge your disk cache before running tests on your end, so that doesn't skew the results?

Purging pycache obviously makes the program slower. That is the nature of python

**Comment by forthrin** [comment]
> > Now, given proper knowledge of the `format` keywords, we should be able to narrow down to the single media file the user is interested in (or is considered most valuable by default)
> 
> How?

I'm not an expert on these formats, but the first to go is mpd with its awful verbose XML. `lb`is low bandwidth, and `CMAF` is a modern replacement for `TS`. So `hls-cmaf-full.m3u8` seems the best candidate for the good stuff.

> The whole startup took only 0.6s this time, not 2.2s like your original run. This much is to be expected since this is a large python program

I realise now I was supposed to run this with the full command line options :/ See [new output file](https://github.com/yt-dlp/yt-dlp/files/10262610/cProfile.txt).

```
sudo purge && time python3 -m cProfile yt_dlp/__main__.py --use-extractor SVTPlay --ignore-config -g https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul
0m4.089s
```

> > PS! Can you purge your disk cache before running tests on your end, so that doesn't skew the results?
> 
> Purging pycache obviously makes the program slower. That is the nature of python

Not pycache, but your disk cache, so that test results from a code change don't appear misleadingly faster, when it's just because stuff resides in memory from the previous run. From the man pages:

```
purge – force disk cache to be purged (flushed and emptied)
Purge can be used to approximate initial boot conditions
with a cold disk buffer cache for performance analysis.
```

**Comment by Jules-A** [comment]
Seems like downloading  site info/m3u8/MPD in parallel would give the biggest speedup but I'd imagine some sites would hate that.

**Comment by forthrin** [comment]
@Jules-A: All these files seem redundant for tech savvy users who use CLI tools. Personally, I always want either the best video quality, or the smallest file/fastest download.

If downloading a large file that takes several minutes, an overhead of 4 seconds is obviously negligible. But when *playing* a file, any delay past a fraction of a second deposits layers of annoyance in the shoulders and cerebral cortex.

Are there any options to prevent fetching all m3u8/MPD? At least immediate attempts like this don't work ...

```
$ yt-dlp -v -f 'bestvideo[protocol*=m3u8]' -g https://www.svtplay.se/video/8MMzVRJ/en-underbar-javla-jul
[SVTPlay] 8MMzVRJ: Downloading MPD manifest
```

**Comment by Jules-A** [comment]
> @Jules-A: All these files seem redundant for tech savvy users who use CLI tools. Personally, I always want either the best video quality, or the smallest file/fastest download.
> 
> If downloading a large file that takes several minutes, an overhead of 4 seconds is obviously negligible. But when _playing_ a file, any delay past a fraction of a second deposits layers of annoyance in the shoulders and cerebral cortex.

Then don't use YT-DLP for playing the files? Create another CLI tool for checking if the file has been downloaded before and play it if so before running YT-DLP...

**Comment by forthrin** [comment]
@Jules-A: Are we talking about the same thing? I'm talking about playing a (any) video, that obviously hasn't been downloaded before, eg. `mpv $(yt-dlp -g $URL)`. This involves a lot of overhead delays, so yes, I created my own scripts for the sites I use most. Well, if `yt-dlp` is meant to be a *download* tool only, and not for instant playing of videos, you are probably right, but I assume it's used for both. I hope significant optimizations can be made in either case.

**Comment by forthrin** [comment]
Getting nowhere slow.
